Mathlib/Order/Heyting/Hom.lean

```diff
@@ -0,0 +1,632 @@
+  { ‹CoheytingHomClass F α β› with
+    map_bot := fun f => by rw [← @sdiff_self α _ ⊤, ← sdiff_self, map_sdiff] }
+#align coheyting_hom_class.to_bounded_lattice_hom_class CoheytingHomClass.toBoundedLatticeHomClass
+
+-- See note [lower instance priority]
+instance (priority := 100) BiheytingHomClass.toHeytingHomClass {_ : BiheytingAlgebra α}
+    { _ : BiheytingAlgebra β} [BiheytingHomClass F α β] : HeytingHomClass F α β :=
+  { ‹BiheytingHomClass F α β› with
+    map_bot := fun f => by rw [← @sdiff_self α _ ⊤, ← sdiff_self, BiheytingHomClass.map_sdiff] }
+#align biheyting_hom_class.to_heyting_hom_class BiheytingHomClass.toHeytingHomClass
+
+-- See note [lower instance priority]
+instance (priority := 100) BiheytingHomClass.toCoheytingHomClass {_ : BiheytingAlgebra α}
+    { _ : BiheytingAlgebra β}  [BiheytingHomClass F α β] : CoheytingHomClass F α β :=
+  { ‹BiheytingHomClass F α β› with
+    map_top := fun f => by rw [← @himp_self α _ ⊥, ← himp_self, map_himp] }
+#align biheyting_hom_class.to_coheyting_hom_class BiheytingHomClass.toCoheytingHomClass
+
+-- See note [lower instance priority]
+instance (priority := 100) OrderIsoClass.toHeytingHomClass {_ : HeytingAlgebra α}
+    { _ : HeytingAlgebra β} [OrderIsoClass F α β] : HeytingHomClass F α β :=
+  { OrderIsoClass.toBoundedLatticeHomClass with
+    map_himp := fun f a b =>
+      eq_of_forall_le_iff fun c =>
+        by
+        simp only [← map_inv_le_iff, le_himp_iff]
+        rw [← OrderIsoClass.map_le_map_iff f]
+        simp }
+#align order_iso_class.to_heyting_hom_class OrderIsoClass.toHeytingHomClass
+
+-- See note [lower instance priority]
+instance (priority := 100) OrderIsoClass.toCoheytingHomClass {_ : CoheytingAlgebra α}
+    { _ : CoheytingAlgebra β} [OrderIsoClass F α β] : CoheytingHomClass F α β :=
+  { OrderIsoClass.toBoundedLatticeHomClass with
+    map_sdiff := fun f a b =>
+      eq_of_forall_ge_iff fun c =>
+        by
+        simp only [← le_map_inv_iff, sdiff_le_iff]
+        rw [← OrderIsoClass.map_le_map_iff f]
+        simp }
+#align order_iso_class.to_coheyting_hom_class OrderIsoClass.toCoheytingHomClass
+
+-- See note [lower instance priority]
+instance (priority := 100) OrderIsoClass.toBiheytingHomClass {_ : BiheytingAlgebra α}
+    { _ : BiheytingAlgebra β} [OrderIsoClass F α β] : BiheytingHomClass F α β :=
+  {
+    OrderIsoClass.toLatticeHomClass with
+    map_himp := fun f a b =>
+      eq_of_forall_le_iff fun c =>
+        by
+        simp only [← map_inv_le_iff, le_himp_iff]
+        rw [← OrderIsoClass.map_le_map_iff f]
+        simp
+    map_sdiff := fun f a b =>
+      eq_of_forall_ge_iff fun c =>
+        by
+        simp only [← le_map_inv_iff, sdiff_le_iff]
+        rw [← OrderIsoClass.map_le_map_iff f]
+        simp }
+#align order_iso_class.to_biheyting_hom_class OrderIsoClass.toBiheytingHomClass
+
+-- See note [reducible non instances]
+/-- This can't be an instance because of typeclass loops. -/
```

Comment:
```suggestion
-- See note [reducible non instances]
/- This can't be an instance because of typeclass loops.
Porting note: Revisit this issue to see if it works in Lean 4. -/
```